File: backend/app/services/video_service.py
from typing import Optional, List
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import func

from app.models import Video, Metadata, Tag, VideoTag, VideoProduction
from app.schemas.video import VideoUpdate, BulkUpdateRequest


class VideoService:
    """Service for video CRUD operations and metadata management."""

    def get_video(self, db: Session, video_id: int) -> Optional[Video]:
        """
        Get a video by ID with all related data.

        Args:
            db: Database session
            video_id: Video ID

        Returns:
            Video object with metadata and tags, or None if not found
        """
        return (
            db.query(Video)
            .options(
                joinedload(Video.video_metadata),
                joinedload(Video.video_tags).joinedload(VideoTag.tag),
                joinedload(Video.video_productions).joinedload(VideoProduction.production)
            )
            .filter(Video.id == video_id)
            .first()
        )

    def get_video_by_path(self, db: Session, file_path: str) -> Optional[Video]:
        """
        Get a video by file path.

        Args:
            db: Database session
            file_path: Video file path

        Returns:
            Video object or None if not found
        """
        return db.query(Video).filter(Video.file_path == file_path).first()

    def update_video(
        self,
        db: Session,
        video_id: int,
        update_data: VideoUpdate
    ) -> Optional[Video]:
        """
        Update video metadata and tags.

        Args:
            db: Database session
            video_id: Video ID
            update_data: Update data

        Returns:
            Updated video object, or None if not found
        """
        video = self.get_video(db, video_id)
        if not video:
            return None

        # Update metadata
        if video.video_metadata is None:
            video.video_metadata = Metadata(video_id=video_id)
            db.add(video.video_metadata)

        if update_data.category is not None:
            video.video_metadata.category = update_data.category
        if update_data.location is not None:
            video.video_metadata.location = update_data.location
        if update_data.notes is not None:
            video.video_metadata.notes = update_data.notes

        # Update tags
        if update_data.tags is not None:
            self._update_video_tags(db, video, update_data.tags)

        # Update productions
        if update_data.production_ids is not None:
            self._update_video_productions(db, video, update_data.production_ids)

        db.commit()

        # Re-fetch to ensure all relationships are fresh
        return self.get_video(db, video_id)

    def bulk_update_videos(
        self,
        db: Session,
        bulk_update: BulkUpdateRequest
    ) -> int:
        """
        Update multiple videos at once.

        Args:
            db: Database session
            bulk_update: Bulk update data

        Returns:
            Number of videos updated
        """
        videos = (
            db.query(Video)
            .options(
                joinedload(Video.video_metadata),
                joinedload(Video.video_tags),
                joinedload(Video.video_productions),
            )
            .filter(Video.id.in_(bulk_update.video_ids))
            .all()
        )

        for video in videos:
            # Ensure metadata exists
            if video.video_metadata is None:
                video.video_metadata = Metadata(video_id=video.id)
                db.add(video.video_metadata)

            # Update metadata fields
            if bulk_update.category is not None:
                video.video_metadata.category = bulk_update.category
            if bulk_update.location is not None:
                video.video_metadata.location = bulk_update.location
            if bulk_update.notes is not None:
                video.video_metadata.notes = bulk_update.notes

            # Add tags
            if bulk_update.add_tags:
                self._add_tags_to_video(db, video, bulk_update.add_tags)

            # Remove tags
            if bulk_update.remove_tags:
                self._remove_tags_from_video(db, video, bulk_update.remove_tags)

            # Add productions
            if bulk_update.add_production_ids:
                self._add_productions_to_video(db, video, bulk_update.add_production_ids)

            # Remove productions
            if bulk_update.remove_production_ids:
                self._remove_productions_from_video(db, video, bulk_update.remove_production_ids)

        db.commit()
        return len(videos)

    def delete_video(self, db: Session, video_id: int) -> bool:
        """
        Delete a video from the database.

        Args:
            db: Database session
            video_id: Video ID

        Returns:
            True if deleted, False if not found
        """
        video = db.query(Video).filter(Video.id == video_id).first()
        if not video:
            return False

        db.delete(video)
        db.commit()
        return True

    def _update_video_tags(self, db: Session, video: Video, tag_names: List[str]) -> None:
        """
        Replace all tags for a video.

        Args:
            db: Database session
            video: Video object
            tag_names: List of tag names
        """
        # Remove existing tags
        db.query(VideoTag).filter(VideoTag.video_id == video.id).delete()

        # Add new tags
        for tag_name in tag_names:
            tag = self._get_or_create_tag(db, tag_name.strip())
            video_tag = VideoTag(video_id=video.id, tag_id=tag.id)
            db.add(video_tag)

    def _add_tags_to_video(self, db: Session, video: Video, tag_names: List[str]) -> None:
        """
        Add tags to a video (doesn't remove existing tags).

        Args:
            db: Database session
            video: Video object
            tag_names: List of tag names to add
        """
        # Get existing tag IDs
        existing_tag_ids = {vt.tag_id for vt in video.video_tags}

        for tag_name in tag_names:
            tag = self._get_or_create_tag(db, tag_name.strip())
            if tag.id not in existing_tag_ids:
                video_tag = VideoTag(video_id=video.id, tag_id=tag.id)
                db.add(video_tag)

    def _remove_tags_from_video(self, db: Session, video: Video, tag_names: List[str]) -> None:
        """
        Remove tags from a video.

        Args:
            db: Database session
            video: Video object
            tag_names: List of tag names to remove
        """
        # Get tag IDs to remove
        tags_to_remove = (
            db.query(Tag.id)
            .filter(Tag.name.in_([t.strip() for t in tag_names]))
            .all()
        )
        tag_ids_to_remove = {t.id for t in tags_to_remove}

        # Remove video_tags
        db.query(VideoTag).filter(
            VideoTag.video_id == video.id,
            VideoTag.tag_id.in_(tag_ids_to_remove)
        ).delete(synchronize_session=False)

    def _get_or_create_tag(self, db: Session, tag_name: str) -> Tag:
        """
        Get an existing tag or create a new one.

        Args:
            db: Database session
            tag_name: Tag name

        Returns:
            Tag object
        """
        tag = db.query(Tag).filter(Tag.name == tag_name).first()
        if not tag:
            tag = Tag(name=tag_name)
            db.add(tag)
            db.flush()
        return tag

    def _update_video_productions(self, db: Session, video: Video, production_ids: List[int]) -> None:
        """
        Replace all production links for a video.

        Args:
            db: Database session
            video: Video object
            production_ids: List of production IDs to link
        """
        # Remove existing production links
        db.query(VideoProduction).filter(VideoProduction.video_id == video.id).delete()

        # Add new production links
        for production_id in production_ids:
            vp = VideoProduction(video_id=video.id, production_id=production_id)
            db.add(vp)

    def _add_productions_to_video(self, db: Session, video: Video, production_ids: List[int]) -> None:
        """Add productions to a video (doesn't remove existing)."""
        existing_prod_ids = {vp.production_id for vp in video.video_productions}
        for prod_id in production_ids:
            if prod_id not in existing_prod_ids:
                vp = VideoProduction(video_id=video.id, production_id=prod_id)
                db.add(vp)

    def _remove_productions_from_video(self, db: Session, video: Video, production_ids: List[int]) -> None:
        """Remove productions from a video."""
        db.query(VideoProduction).filter(
            VideoProduction.video_id == video.id,
            VideoProduction.production_id.in_(production_ids)
        ).delete(synchronize_session=False)

    def get_all_categories(self, db: Session) -> List[dict]:
        """
        Get all unique categories with usage count.

        Args:
            db: Database session

        Returns:
            List of category dictionaries with name and count
        """
        results = (
            db.query(
                Metadata.category,
                func.count(Metadata.category).label('count')
            )
            .filter(Metadata.category.isnot(None))
            .group_by(Metadata.category)
            .order_by(Metadata.category)
            .all()
        )

        return [{"name": cat, "count": count} for cat, count in results]

    def get_all_tags(self, db: Session) -> List[dict]:
        """
        Get all tags with usage count.

        Args:
            db: Database session

        Returns:
            List of tag dictionaries with id, name, and count
        """
        results = (
            db.query(
                Tag.id,
                Tag.name,
                func.count(VideoTag.video_id).label('count')
            )
            .outerjoin(VideoTag, Tag.id == VideoTag.tag_id)
            .group_by(Tag.id, Tag.name)
            .order_by(Tag.name)
            .all()
        )

        return [{"id": id, "name": name, "count": count} for id, name, count in results]


# Global video service instance
video_service = VideoService()

File: backend-rust/src/services/video_service.rs
use diesel::prelude::*;
use diesel::sqlite::SqliteConnection;

use crate::models::*;
use crate::schema::{videos, metadata, productions, tags, video_tags, video_productions};

/// Get a single video with all relationships
pub fn get_video(conn: &mut SqliteConnection, video_id: i32) -> Option<VideoFull> {
    let video: Video = videos::table
        .find(video_id)
        .first(conn)
        .ok()?;

    Some(load_video_full(conn, video))
}

#[allow(dead_code)]
/// Get a video by its file path
pub fn get_video_by_path(conn: &mut SqliteConnection, file_path: &str) -> Option<Video> {
    videos::table
        .filter(videos::file_path.eq(file_path))
        .first(conn)
        .ok()
}

/// Load full video data with metadata, tags, and productions
pub fn load_video_full(conn: &mut SqliteConnection, video: Video) -> VideoFull {
    let meta: Option<Metadata> = metadata::table
        .filter(metadata::video_id.eq(video.id))
        .first(conn)
        .ok();

    let tag_list: Vec<Tag> = video_tags::table
        .inner_join(tags::table)
        .filter(video_tags::video_id.eq(video.id))
        .select(tags::all_columns)
        .load(conn)
        .unwrap_or_default();

    let prod_list: Vec<Production> = video_productions::table
        .inner_join(crate::schema::productions::table)
        .filter(video_productions::video_id.eq(video.id))
        .select(crate::schema::productions::all_columns)
        .load(conn)
        .unwrap_or_default();

    let orientation = orientation_from_resolution(&video.resolution);

    VideoFull {
        id: video.id,
        file_path: video.file_path,
        filename: video.filename,
        duration: video.duration,
        file_size: video.file_size,
        resolution: video.resolution,
        fps: video.fps,
        codec: video.codec,
        created_date: video.created_date,
        indexed_date: video.indexed_date,
        thumbnail_count: video.thumbnail_count,
        checksum: video.checksum,
        orientation,
        metadata: meta.map(|m| MetadataResponse {
            category: m.category,
            location: m.location,
            notes: m.notes,
        }),
        tags: tag_list.into_iter().map(|t| TagResponse {
            id: t.id,
            name: t.name,
        }).collect(),
        productions: prod_list.into_iter().map(|p| ProductionBriefResponse {
            id: p.id,
            title: p.title,
            platform: p.platform,
            link: p.link,
            is_published: p.is_published,
        }).collect(),
    }
}

/// Load full data for multiple videos
pub fn load_videos_full(conn: &mut SqliteConnection, video_list: Vec<Video>) -> Vec<VideoFull> {
    video_list.into_iter()
        .map(|v| load_video_full(conn, v))
        .collect()
}

/// Update a video's metadata, tags, and productions
pub fn update_video(
    conn: &mut SqliteConnection,
    video_id: i32,
    update: &VideoUpdate,
) -> Option<VideoFull> {
    // Verify video exists
    let _video: Video = videos::table.find(video_id).first(conn).ok()?;

    // Ensure metadata record exists
    ensure_metadata_exists(conn, video_id);

    // Update metadata fields
    if update.category.is_some() || update.location.is_some() || update.notes.is_some() {
        let changeset = MetadataChangeset {
            category: update.category.clone(),
            location: update.location.clone(),
            notes: update.notes.clone(),
        };
        diesel::update(metadata::table.filter(metadata::video_id.eq(video_id)))
            .set(&changeset)
            .execute(conn)
            .ok();
    }

    // Replace tags if provided
    if let Some(ref tag_names) = update.tags {
        update_video_tags(conn, video_id, tag_names);
    }

    // Replace productions if provided
    if let Some(ref prod_ids) = update.production_ids {
        update_video_productions(conn, video_id, prod_ids);
    }

    get_video(conn, video_id)
}

/// Why a video could not be deleted.
pub enum DeleteVideoError {
    NotFound,
    /// The video is linked to one or more productions (their titles).
    UsedInProductions(Vec<String>),
    Db(String),
}

/// What happened on disk during a successful delete.
pub struct DeleteVideoOutcome {
    /// True when the source file was removed (or was already gone).
    pub file_deleted: bool,
    /// Set when file removal was requested but failed (e.g. file in use).
    pub file_error: Option<String>,
}

/// Delete a video from the library. Refuses when the video is used in any
/// production — those are part of an edit history and must be unlinked first.
/// DB relationships cascade; the thumbnail folder is removed unless a
/// duplicate video shares the same checksum. With `delete_file`, the source
/// file is also removed from disk (best-effort — the DB row is already gone,
/// so a locked file is reported rather than rolling back).
pub fn delete_video_checked(
    conn: &mut SqliteConnection,
    video_id: i32,
    delete_file: bool,
    thumbnail_dir: &std::path::Path,
) -> Result<DeleteVideoOutcome, DeleteVideoError> {
    let video: Video = videos::table
        .find(video_id)
        .first(conn)
        .map_err(|_| DeleteVideoError::NotFound)?;

    let used_in: Vec<String> = video_productions::table
        .inner_join(productions::table)
        .filter(video_productions::video_id.eq(video_id))
        .select(productions::title)
        .load(conn)
        .map_err(|e| DeleteVideoError::Db(e.to_string()))?;
    if !used_in.is_empty() {
        return Err(DeleteVideoError::UsedInProductions(used_in));
    }

    diesel::delete(videos::table.find(video_id))
        .execute(conn)
        .map_err(|e| DeleteVideoError::Db(e.to_string()))?;

    // Thumbnails are stored per-checksum; only remove the folder when no
    // other (duplicate) video still points at it.
    if let Some(checksum) = video.checksum.as_deref().filter(|c| !c.is_empty()) {
        let still_shared: i64 = videos::table
            .filter(videos::checksum.eq(checksum))
            .count()
            .get_result(conn)
            .unwrap_or(0);
        if still_shared == 0 {
            let _ = std::fs::remove_dir_all(thumbnail_dir.join(checksum));
        }
    }

    let mut outcome = DeleteVideoOutcome { file_deleted: false, file_error: None };
    if delete_file {
        let path = std::path::Path::new(&video.file_path);
        if !path.exists() {
            outcome.file_deleted = true; // already gone from disk
        } else {
            match std::fs::remove_file(path) {
                Ok(()) => outcome.file_deleted = true,
                Err(e) => outcome.file_error = Some(e.to_string()),
            }
        }
    }
    Ok(outcome)
}

/// Bulk update multiple videos
pub fn bulk_update_videos(
    conn: &mut SqliteConnection,
    bulk: &BulkUpdateRequest,
) -> Result<i64, String> {
    let video_list: Vec<Video> = videos::table
        .filter(videos::id.eq_any(&bulk.video_ids))
        .load(conn)
        .map_err(|e| e.to_string())?;

    let count = video_list.len() as i64;

    for video in &video_list {
        ensure_metadata_exists(conn, video.id);

        // Update fields
        if bulk.category.is_some() || bulk.location.is_some() || bulk.notes.is_some() {
            let changeset = MetadataChangeset {
                category: bulk.category.clone(),
                location: bulk.location.clone(),
                notes: bulk.notes.clone(),
            };
            diesel::update(metadata::table.filter(metadata::video_id.eq(video.id)))
                .set(&changeset)
                .execute(conn)
                .ok();
        }

        // Add tags
        if let Some(ref tag_names) = bulk.add_tags {
            add_tags_to_video(conn, video.id, tag_names);
        }

        // Remove tags
        if let Some(ref tag_names) = bulk.remove_tags {
            remove_tags_from_video(conn, video.id, tag_names);
        }

        // Add productions
        if let Some(ref prod_ids) = bulk.add_production_ids {
            add_productions_to_video(conn, video.id, prod_ids);
        }

        // Remove productions
        if let Some(ref prod_ids) = bulk.remove_production_ids {
            remove_productions_from_video(conn, video.id, prod_ids);
        }
    }

    Ok(count)
}

// --- Tag helpers ---

fn get_or_create_tag(conn: &mut SqliteConnection, tag_name: &str) -> i32 {
    if let Ok(tag) = tags::table
        .filter(tags::name.eq(tag_name))
        .first::<Tag>(conn)
    {
        return tag.id;
    }

    diesel::insert_into(tags::table)
        .values(&NewTag { name: tag_name.to_string() })
        .execute(conn)
        .ok();

    tags::table
        .filter(tags::name.eq(tag_name))
        .first::<Tag>(conn)
        .map(|t| t.id)
        .unwrap_or(0)
}

fn update_video_tags(conn: &mut SqliteConnection, video_id: i32, tag_names: &[String]) {
    // Delete all existing tags for this video
    diesel::delete(video_tags::table.filter(video_tags::video_id.eq(video_id)))
        .execute(conn)
        .ok();

    // Add new tags
    for name in tag_names {
        let tag_id = get_or_create_tag(conn, name);
        diesel::insert_into(video_tags::table)
            .values(&NewVideoTag { video_id, tag_id })
            .execute(conn)
            .ok();
    }
}

fn add_tags_to_video(conn: &mut SqliteConnection, video_id: i32, tag_names: &[String]) {
    let existing_tag_ids: Vec<i32> = video_tags::table
        .filter(video_tags::video_id.eq(video_id))
        .select(video_tags::tag_id)
        .load(conn)
        .unwrap_or_default();

    for name in tag_names {
        let tag_id = get_or_create_tag(conn, name);
        if !existing_tag_ids.contains(&tag_id) {
            diesel::insert_into(video_tags::table)
                .values(&NewVideoTag { video_id, tag_id })
                .execute(conn)
                .ok();
        }
    }
}

fn remove_tags_from_video(conn: &mut SqliteConnection, video_id: i32, tag_names: &[String]) {
    let tag_ids: Vec<i32> = tags::table
        .filter(tags::name.eq_any(tag_names))
        .select(tags::id)
        .load(conn)
        .unwrap_or_default();

    diesel::delete(
        video_tags::table
            .filter(video_tags::video_id.eq(video_id))
            .filter(video_tags::tag_id.eq_any(&tag_ids)),
    )
    .execute(conn)
    .ok();
}

// --- Production helpers ---

fn update_video_productions(conn: &mut SqliteConnection, video_id: i32, prod_ids: &[i32]) {
    diesel::delete(video_productions::table.filter(video_productions::video_id.eq(video_id)))
        .execute(conn)
        .ok();

    for &prod_id in prod_ids {
        diesel::insert_into(video_productions::table)
            .values(&NewVideoProduction { video_id, production_id: prod_id })
            .execute(conn)
            .ok();
    }
}

fn add_productions_to_video(conn: &mut SqliteConnection, video_id: i32, prod_ids: &[i32]) {
    let existing: Vec<i32> = video_productions::table
        .filter(video_productions::video_id.eq(video_id))
        .select(video_productions::production_id)
        .load(conn)
        .unwrap_or_default();

    for &prod_id in prod_ids {
        if !existing.contains(&prod_id) {
            diesel::insert_into(video_productions::table)
                .values(&NewVideoProduction { video_id, production_id: prod_id })
                .execute(conn)
                .ok();
        }
    }
}

fn remove_productions_from_video(conn: &mut SqliteConnection, video_id: i32, prod_ids: &[i32]) {
    diesel::delete(
        video_productions::table
            .filter(video_productions::video_id.eq(video_id))
            .filter(video_productions::production_id.eq_any(prod_ids)),
    )
    .execute(conn)
    .ok();
}

// --- Metadata helpers ---

fn ensure_metadata_exists(conn: &mut SqliteConnection, video_id: i32) {
    let exists = metadata::table
        .filter(metadata::video_id.eq(video_id))
        .count()
        .get_result::<i64>(conn)
        .unwrap_or(0) > 0;

    if !exists {
        diesel::insert_into(metadata::table)
            .values(&NewMetadata {
                video_id,
                category: None,
                location: None,
                notes: None,
            })
            .execute(conn)
            .ok();
    }
}
